ddl/ddl_api.go

```diff
@@ -5177,62 +5233,69 @@ func checkPlacementSpecs(specs []*ast.PlacementSpec) ([]*placement.Rule, error)
 		switch spec.Tp {
 		case ast.PlacementAdd:
 		default:
-			return rules, errors.Errorf("invalid placement spec[%d], unknown action type: %d", k, spec.Tp)
-		}
-
-		switch spec.Role {
-		case ast.PlacementRoleFollower:
-			rule.Role = placement.Follower
-		case ast.PlacementRoleLeader:
-			rule.Role = placement.Leader
-		case ast.PlacementRoleLearner:
-			rule.Role = placement.Learner
-		case ast.PlacementRoleVoter:
-			rule.Role = placement.Voter
-		default:
-			return rules, errors.Errorf("invalid placement spec[%d], unknown role: %d", k, spec.Role)
-		}
-
-		for _, label := range strings.Split(spec.Constraints, ",") {
-			label = strings.TrimSpace(label)
-
-			if len(label) < 4 {
-				return rules, errors.Errorf("invalid placement spec[%d], constraint too short to be valid: %s", k, label)
-			}
-
-			var op placement.LabelConstraintOp
-			switch label[0] {
-			case '+':
-				op = placement.In
-			case '-':
-				op = placement.NotIn
+			err = errors.Errorf("unknown action type: %d", spec.Tp)
+		}
+
+		if err == nil {
+			switch spec.Role {
+			case ast.PlacementRoleFollower:
+				rule.Role = placement.Follower
+			case ast.PlacementRoleLeader:
+				rule.Role = placement.Leader
+			case ast.PlacementRoleLearner:
+				rule.Role = placement.Learner
+			case ast.PlacementRoleVoter:
+				rule.Role = placement.Voter
 			default:
-				return rules, errors.Errorf("invalid placement spec[%d], unknown operation: %c", k, label[0])
+				err = errors.Errorf("unknown role: %d", spec.Role)
 			}
+		}
 
-			kv := strings.Split(label[1:], "=")
-			if len(kv) != 2 {
-				return rules, errors.Errorf("invalid placement spec[%d], invalid constraint format: %s", k, label)
-			}
+		if err == nil {
+			cnstr := strings.TrimSpace(spec.Constraints)
 
-			key := strings.TrimSpace(kv[0])
-			if key == "" {
-				return rules, errors.Errorf("invalid placement spec[%d], empty constraint key: %s", k, label)
+			if len(cnstr) > 0 && cnstr[0] == '[' {
+				constraints := []string{}
+				err = json.Unmarshal([]byte(cnstr), &constraints)
+				if err == nil {
+					err = checkPlacementLabelConstraints(rule, constraints)
+					if err == nil {
+						rules = append(rules, rule)
+					}
+				}
+			} else if len(cnstr) > 0 && cnstr[0] == '{' {
+				constraints := map[string]int{}
+				err = json.Unmarshal([]byte(cnstr), &constraints)
+				if err == nil {
+					for labels, cnt := range constraints {
+						newrule := &placement.Rule{}
+						*newrule = *rule
+						if cnt <= 0 {
```

Comment:
This is too much nesting, could we extract these code(line5225-5289) into a function?

Reply:
5255? Anyway, it is  better now.